core/src/main/java/org/infinispan/cache/impl/EncoderCache.java

```diff
@@ -221,6 +229,29 @@ public boolean contains(Object o) {
       public boolean remove(Object o) {
          return actualCollection.remove(keyToStorage(o));
       }
+
+      // These methods should be delegated to actual collection as they can be faster
+
+      @Override
+      public boolean removeAll(Collection<?> c) {
+         return actualCollection.removeAll(buildKeyCollection(c));
+      }
+
+      @Override
+      public boolean retainAll(Collection<?> c) {
+         return actualCollection.retainAll(buildKeyCollection(c));
+      }
+
+      Collection<K> buildKeyCollection(Collection<?> c) {
+         Collection<K> toRemove = new ArrayList<>(c.size());
+         for (Object obj : c) {
+            K entry = keyToStorage(obj);
```

Comment:
So I was debating about both these points a bit.

I think I need to go back and just implement all the methods and let the underlying sets do their jobs. I should make a JIRA to do this though since it isn't directly related to the future changes here.

Reply:
I created https://issues.jboss.org/browse/ISPN-8946